processing/src/main/java/io/druid/query/dimension/LookupDimensionSpec.java

```diff
@@ -0,0 +1,206 @@
+/*
+ * Licensed to Metamarkets Group Inc. (Metamarkets) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. Metamarkets licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.query.dimension;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.common.base.Preconditions;
+import com.google.common.base.Strings;
+import com.metamx.common.StringUtils;
+import io.druid.query.extraction.DimExtractionFn;
+import io.druid.query.extraction.ExtractionFn;
+import io.druid.query.extraction.LookupExtractor;
+import io.druid.query.filter.DimFilterCacheHelper;
+import io.druid.segment.DimensionSelector;
+
+import java.nio.ByteBuffer;
+
+public class LookupDimensionSpec implements DimensionSpec
+{
+  private static final byte CACHE_TYPE_ID = 0x4;
+
+  @JsonProperty
+  private final String dimension;
+
+  @JsonProperty
+  private final String outputName;
+
+  @JsonProperty
+  private final LookupExtractor lookup;
+
+  @JsonProperty
+  private final boolean retainMissingValues;
+
+  @JsonProperty
+  private final String replaceMissingWith;
+
+  private final ExtractionFn extractionFn;
+
+  @JsonCreator
+  public LookupDimensionSpec(
```

Comment:
What do you think about making this

```
public LookupDimensionSpec(
  @JsonProperty("dimension") String dimension,
  @JsonProperty("outputName") String outputName,
  @JsonProperty("lookup") LookupExtractor lookupInput,
  @JsonProperty("name") String name,
  @JsonProperty("retainMissingValues") final boolean retainMissingValues,
  @JsonProperty("replaceMissingWith") final String replaceMissingWith,
  @JacksonInject LookupReferencesManager lookupReferencesManager,
)
```

Where it is expected that you either specify "name" or "lookup", the one you do not specify is null.

If you specify lookup, then said lookup is used when `getExtractionFn()` is called.  If you specify name, then said name is looked up in the manager when `getExtractionFn()` is called.

Structuring it this way should avoid issues with the lookup not existing on the router and completely eliminate the need for the `LookupDelegator`